[Lens] Add Multi terms support to Top Values

dej611 opened the pull request:
## Summary

Fix #95079

This PR introduces an interface for Lens of the Multi terms value aggregation, via the Top Values function.
This PR does not contain the `Combine` drag and drop action. Will be added in a follow up PR.

When a new term field is added the current fields used are removed from the field dropdown list.
When more than one field is defined `scripted fields` are filtered from the field dropdown list, and where already set are marked as invalid with an explanation (both in the workspace and in the panel) on how to fix it.

<img width="974" alt="Screenshot 2021-11-25 at 15 14 32" src="https://user-images.githubusercontent.com/924948/143457503-51c56b80-b36e-4c7d-829b-666b14c9271f.png">

<img width="968" alt="Screenshot 2021-11-25 at 15 14 44" src="https://user-images.githubusercontent.com/924948/143457517-13ef8739-d66d-4d26-9980-e902e38bc7f5.png">

<img width="983" alt="Screenshot 2021-11-25 at 15 15 24" src="https://user-images.githubusercontent.com/924948/143457535-034fb2cc-332a-40d7-aff9-cc33499c3675.png">

<img width="971" alt="Screenshot 2021-11-25 at 15 15 39" src="https://user-images.githubusercontent.com/924948/143457552-1c898261-024c-4dc0-a93e-61d360249470.png">

Time shift conflict is resolved similarly for the single term, concatenating conditions of multiple terms:

<img width="967" alt="Screenshot 2021-11-25 at 15 16 00" src="https://user-images.githubusercontent.com/924948/143457645-cdf57298-ef1f-41c1-8207-e554380aaa4a.png">
<img width="960" alt="Screenshot 2021-11-25 at 15 16 14" src="https://user-images.githubusercontent.com/924948/143457676-0d602778-3718-4233-84dd-f1a71c5e8670.png">
<img width="988" alt="Screenshot 2021-11-25 at 15 16 31" src="https://user-images.githubusercontent.com/924948/143457703-d0749d62-a4d1-4773-b830-502749490422.png">

Terms can be reordered manually via Drag and Drop:

![reorder_terms](https://user-images.githubusercontent.com/924948/143457784-4bcd487b-0c81-40ca-85fe-c9d221922d2e.gif)

The empty field can be reordered as well but it has no effect until it contains a value. Closing and reopening the panel when empty will just discard it:

![reorder_empty_terms](https://user-images.githubusercontent.com/924948/143457818-f1fabfe4-303b-444f-bed8-dea5a561ed0e.gif)

In case of 2 fields, where only one is filled with a value, the trash icon will be enabled only for the empty one:
<img width="980" alt="Screenshot 2021-11-25 at 15 21 28" src="https://user-images.githubusercontent.com/924948/143458052-cbdc4019-a16e-47d0-8c3b-be46442632fb.png">

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

ghudgins commented:
Will list things here as I find them: 
 - couldn't filter when left clicking multi terms in a stacked bar, it just filters time. Expected the "Select filters to apply" modal with checkbox choices

dej611 commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

dej611 commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

dej611 commented:
@elasticmachine merge upstream

dej611 commented:
> couldn't filter when left clicking multi terms in a stacked bar, it just filters time. Expected the "Select filters to apply" modal with checkbox choices

This now should have been sorted.

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

dej611 commented:
@elasticmachine merge upstream

mbondyra commented:
@elasticmachine merge upstream

mbondyra commented:
Questions:
1. Should we limit number of  fields user can use?
<img width="299" alt="Screenshot 2021-11-29 at 14 03 56" src="https://user-images.githubusercontent.com/4283304/143873126-f004dd71-04d9-43f2-b27e-77684041ae23.png">

2. A question about filters behavior (Graham already pointed it out, but it still doesn't work correctly for me)
a. When clicking on a chart bar only time is filtered (checkboxes popover doesn't appear)


https://user-images.githubusercontent.com/4283304/143873961-59c9cd78-db39-4075-adae-d8be64aa1386.mp4

b. When clicking on a legend, the filter is being created without the ability to choose which of the fields we want to filter

https://user-images.githubusercontent.com/4283304/143873970-adb0b879-8488-42fc-8788-9eaf4bf5e742.mp4

mbondyra commented on src/plugins/data/common/search/aggs/buckets/multi_terms_fn.ts:
```diff
@@ -111,6 +111,12 @@ export const aggMultiTerms = (): FunctionDefinition => ({
+    separatorLabel: {
+      types: ['string'],
+      help: i18n.translate('data.search.aggs.buckets.multiTerms.separatorLabel.help', {
+        defaultMessage: 'The separator label used to join each term combination',
```
curious: is this exposed to the user? Can we somehow change this separator?

dej611 replied:
ATM the expression is not yet exposed to the final user, but it may be consumed by others, which may use another separator.
I thought that it won't be harmful to expose it as function argument and would provide some flexibility to others.

mbondyra commented on x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/field_input.test.tsx:
```diff
@@ -0,0 +1,487 @@
+    ).toBe('Invalid field. Check your data view or pick another field.');
+  });
+
+  it('should render an help message when passed and no errors are found', () => {
```
nit, typo
```suggestion
  it('should render a help message when passed and no errors are found', () => {
```

flash1293 commented:
I wasn’t aware we wanted to filter this way - right now a multi term value is treated as a single filterable thing. I thought that’s the mental model we are after - is it a common thing to filter by a part of a multi terms value? If yes I guess it’s possible to adjust the filter building logic on app services side to get the modal for this case (but not for “other” because it’s grouping by both “or” and “and” which can’t be represented)

mbondyra commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/terms/field_inputs.tsx:
```diff
@@ -0,0 +1,232 @@
+                (!rawValuesLookup.has(key) && !indexPattern.getFieldByName(key)?.scripted) ||
+                key === value
+            )
+            .reduce((memo, key) => {
+              memo[key] = operationSupportMatrix.operationByField[key];
+              return memo;
+            }, {} as OperationSupportMatrix['operationByField']);
```
to avoid type assertion:
```suggestion
            .reduce<OperationSupportMatrix['operationByField']>((memo, key) => {
              memo[key] = operationSupportMatrix.operationByField[key];
              return memo;
            }, {});
```

mbondyra commented on x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/field_input.tsx:
```diff
@@ -0,0 +1,114 @@
+      selectedOperationDefinition?.input !== 'managedReference') ||
+      (incompleteOperation && operationDefinitionMap[incompleteOperation].input === 'field')) &&
+    getErrorMessage(
+      selectedColumn as GenericIndexPatternColumn,
```
this type assertion shouldn't be needed.
```suggestion
      selectedColumn,
```

dej611 commented:
> Questions:
> 
> 1. Should we limit number of  fields user can use?
> 
> ...

Probably 3 (or 4?) can be enough I believe.

> 
> 2. A question about filters behavior (Graham already pointed it out, but it still doesn't work correctly for me)
>    a. When clicking on a chart bar only time is filtered (checkboxes popover doesn't appear)
> 
>  Nov-29-2021.14-09-40.mp4 
> b. When clicking on a legend, the filter is being created without the ability to choose which of the fields we want to filter
> 
>  Nov-29-2021.14-10-11.mp4 

I see where you are coming from. I think it may make sense, but I'd propose to handle it in a follow up PR: right now the chart doesn't know anything about multi terms.

> 3. For the smaller screens, the items have different width:
> 
> ...
> 
> 5. We should wrap/truncate long names of fields:
> 
> ...

I believe these 2 are related. Looking right now for a fix.

> 4. I agree with @ghudgins that we should change truncating setting to not truncate for multiterms.

Also in this case I think it makes sense. Will start to make the chart aware of the multi terms function here, and later on it could be extended for 2.

dej611 commented:
@elasticmachine merge upstream

dej611 commented:
Pushed fixes for 1, 3, 4, 5 issues raised.
I'd leave the 2 as a follow up PR if it's ok.

dej611 commented:
Sorted out the filtering issue with multi terms raised by @ghudgins .

Dosant commented on src/plugins/data/common/search/aggs/utils/get_aggs_formats.ts:
```diff
@@ -143,9 +143,11 @@ export function getAggsFormats(getFieldFormat: GetFieldFormat): FieldFormatInsta
           return params.otherBucketLabel;
         }
 
+        const joinTemplate = ` ${params.separator ?? '-'} `;
```
We changed default from ` > ` to `-` is this expected?

dej611 replied:
I think this change happened before this commit who change the default from `-` to `>`: https://github.com/elastic/kibana/pull/116928/commits/9ef5e19573a41dfbd279ab8b1c0f10afe0f9531d

I can set `>` as default. Do you think it is still worth leaving the option?

Dosant replied:
I'd leave the option and keep ` > ` as default (note whitespaces around it)

Dosant reviewed (approved):
App service agg change lgtm, 

Would be great to add a unit test for `AggsMultiTermsFieldFormat`, we have it for other formats in that file 

And there is a question about default separator:

mbondyra commented:
The issues I mentioned before are all fixed. One small thing to correct (here or in the next PR if you prefer to:)
* dropping a filed to 2-fields multifield replaces the first field. Not sure if that's desired behavior. Maybe instead we should replace the whole dimension with a new top values? 

https://user-images.githubusercontent.com/4283304/144404158-171b8392-16f4-4c56-841e-06818c57f31e.mp4


* when switching between top values and filters we make a default filter with the field from top values. For multi values, we only do it for the first field. Maybe it's worth to expand it to all fields?

https://user-images.githubusercontent.com/4283304/144404173-5ada7fe5-bc75-471b-8bf4-3b41962e5941.mp4

dej611 commented:
> * dropping a filed to 2-fields multifield replaces the first field. Not sure if that's desired behavior. Maybe instead we should replace the whole dimension with a new top values?

The drag and drop behaviour has not be considered in this PR yet, but I have prepared a follow up PR which adds a `combine` action on drop, plus other drop handlers: https://github.com/elastic/kibana/pull/119841
I'll make sure to add a test case on that branch for this replace scenario.

> * when switching between top values and filters we make a default filter with the field from top values. For multi values, we only do it for the first field. Maybe it's worth to expand it to all fields?

This is definitely a bug in this PR. Will address it.

dej611 commented:
> I'd leave the option and keep > as default (note whitespaces around it)
@Dosant I've updated the default char in the formatter.
Right now I've just replaced the single char: do you prefer having a full string replacement instead?
```js
const separator = params.separator ?? ' > '
```

> * dropping a filed to 2-fields multifield replaces the first field. Not sure if that's desired behavior. Maybe instead we should replace the whole dimension with a new top values?

I've tested this behaviour also in th linked PR and it was a full bug. Fixed the behaviour plus one more unit test for it. 👍 

> when switching between top values and filters we make a default filter with the field from top values. For multi values, we only do it for the first field. Maybe it's worth to expand it to all fields?

Also addressed this, adding tests on the filters side.

Dosant commented:
> > I'd leave the option and keep > as default (note whitespaces around it)
> 
> @Dosant I've updated the default char in the formatter. Right now I've just replaced the single char: do you prefer having a full string replacement instead?
> 
> ```js
> const separator = params.separator ?? ' > '
> ```


```
const separator = params.separator ?? ' > '
```

is more flexible, so I'd keep as it was :D

dej611 commented:
@Dosant restored the previous version + added a couple of unit tests.

stratoula commented:
@dej611 while the legend is not truncated for xy charts 
![image](https://user-images.githubusercontent.com/17003240/144613040-67ac2ef8-4e90-4783-9c3a-c3d0dcee571a.png)

for the partition charts is truncated. Is this intended?
![image](https://user-images.githubusercontent.com/17003240/144613419-4b652779-d550-43b2-957d-73826b6ad2c4.png)

stratoula reviewed (approved):
I am approving as I have only found this minor issue. Feel free to merge when this is solved! What a lovely PR 👏

dej611 commented:
@stratoula refactored the truncation logic into a shared helper and added few tests for it.
Now XY, Pie and Heatmap charts use the shared function for truncation default.

flash1293 commented:
@elasticmachine merge upstream

flash1293 commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

mbondyra commented:
@elasticmachine merge upstream
